File: spot-service/src/main/kotlin/it/unibo/harmonikt/repository/SpotMarkerRepository.kt
package it.unibo.harmonikt.repository

import it.unibo.harmonikt.model.Marker.SpotMarker
import kotlin.uuid.Uuid

/**
 * Repository interface for managing SpotMarker entities.
 */
interface SpotMarkerRepository : MarkerRepository<SpotMarker>

/**
 * Fake implementation of SpotMarkerRepository for testing purposes.
 * Stores markers in memory.
 */
class FakeSpotMarkerRepository : SpotMarkerRepository {
    private val markers = mutableListOf<SpotMarker>(
        SpotMarker(Uuid.random(), "Waypoint 1"),
        SpotMarker(Uuid.random(), "Waypoint 2"),
        SpotMarker(Uuid.random(), "Waypoint 3"),
    )

    /**
     * Returns the list of all spot markers.
     */
    override fun getMarkers(): List<SpotMarker> = markers

    /**
     * Returns the spot marker with the given id, or null if not found.
     *
     * @param id the unique identifier of the marker
     */
    override fun getMarkerById(id: Uuid): SpotMarker? = markers.find { it.id == id }

    /**
     * Adds a new spot marker to the repository.
     *
     * @param spotMarker the marker to add
     * @throws IllegalArgumentException if a marker with the same id already exists
     */
    override fun createMarker(spotMarker: SpotMarker) {
        requireNotNull(getMarkerById(spotMarker.id)) { "Marker with id ${spotMarker.id} already exists." }
        markers.add(spotMarker)
    }

    /**
     * Deletes the spot marker with the given id.
     *
     * @param id the unique identifier of the marker to delete
     * @return true if the marker was removed, false otherwise
     */
    override fun deleteMarker(id: Uuid): Boolean = markers.removeIf { it.id == id }
}

File: spot-service-py/app/repositories/spot_marker_repository.py
from abc import ABC, abstractmethod
from typing import List, Optional
from uuid import UUID, uuid4

from app.models import SpotMarker

class MarkerRepository(ABC):
    """
    Abstract base class for marker repositories.
    """
    @abstractmethod
    def get_markers(self) -> List[SpotMarker]:
        """Returns the list of all markers."""
        pass

    @abstractmethod
    def get_marker_by_id(self, marker_id: UUID) -> Optional[SpotMarker]:
        """Returns the marker with the given id, or None if not found."""
        pass

    @abstractmethod
    def create_marker(self, marker: SpotMarker) -> None:
        """Adds a new marker to the repository."""
        pass

    @abstractmethod
    def delete_marker(self, marker_id: UUID) -> bool:
        """Deletes the marker with the given id. Returns True if successful, False otherwise."""
        pass


class FakeSpotMarkerRepository(MarkerRepository):
    """
    Fake implementation of SpotMarkerRepository for testing purposes.
    Stores markers in memory.
    """
    def __init__(self):
        self.markers = [
            SpotMarker(id=uuid4(), waypoint="Waypoint 1"),
            SpotMarker(id=uuid4(), waypoint="Waypoint 2"),
            SpotMarker(id=uuid4(), waypoint="Waypoint 3"),
        ]

    def get_markers(self) -> List[SpotMarker]:
        """Returns the list of all spot markers."""
        return self.markers

    def get_marker_by_id(self, marker_id: UUID) -> Optional[SpotMarker]:
        """Returns the spot marker with the given id, or None if not found."""
        return next((marker for marker in self.markers if marker.id == marker_id), None)

    def create_marker(self, marker: SpotMarker) -> None:
        """
        Adds a new spot marker to the repository.
        Raises ValueError if a marker with the same id already exists.
        """
        if self.get_marker_by_id(marker.id) is not None:
            raise ValueError(f"Marker with id {marker.id} already exists.")
        self.markers.append(marker)

    def delete_marker(self, marker_id: UUID) -> bool:
        """
        Deletes the spot marker with the given id.
        Returns True if the marker was removed, False otherwise.
        """
        marker = self.get_marker_by_id(marker_id)
        if marker:
            self.markers.remove(marker)
            return True
        return False
